### PR TITLE
8339643: Port JEP 404 to RISC-V

### PR DESCRIPTION
**Pull request by RealFYang**
The riscv64 implementation of [JDK-8337511](https://bugs.openjdk.org/browse/JDK-8337511) Implement JEP-404: Generational Shenandoah (Experimental).
This is mainly adapted from the changes in aarch64.

Testing
- [x] Gtest (release & fastdebug)
- [x] hotspot_gc_shenandoah (release & fastdebug)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8339643](https://bugs.openjdk.org/browse/JDK-8339643): Port JEP 404 to RISC-V (**Enhancement** - P4)

### Reviewers
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - Committer) ⚠️ Review applies to [ab573a27](https://git.openjdk.org/shenandoah/pull/493/files/ab573a2720531992ae23d03aefb5477a1fe80e27)
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer) ⚠️ Review applies to [54c5f73b](https://git.openjdk.org/shenandoah/pull/493/files/54c5f73bfbc36289231ddf8494410c53382bd785)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/493/head:pull/493` \
`$ git checkout pull/493`

Update a local copy of the PR: \
`$ git checkout pull/493` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/493/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 493`

View PR using the GUI difftool: \
`$ git pr show -t 493`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/493.diff">https://git.openjdk.org/shenandoah/pull/493.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/shenandoah/pull/493#issuecomment-2334974423)

### CODE REVIEW

**Review by earthling-amzn — APPROVED** [review]
Looks good to me. Thank you!

**Comment by earthling-amzn** [comment]
/reviewers 2

Requesting another reviewer for platform specific code.

**Comment by earthling-amzn** [comment]
Could you please enable the github actions for testing in your fork of this repo?

File: src/hotspot/share/gc/shenandoah/mode/shenandoahGenerationalMode.cpp [hunk 1]
```diff
@@ -32,7 +32,7 @@
 
 void ShenandoahGenerationalMode::initialize_flags() const {
 
-#if !(defined AARCH64 || defined AMD64 || defined IA32 || defined PPC64)
+#if !(defined AARCH64 || defined AMD64 || defined IA32 || defined PPC64 || RISCV64)
   vm_exit_during_initialization("Shenandoah Generational GC is not supported on this platform.");
```
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/mode/shenandoahGenerationalMode.cpp:36]: Should RISCV64 use the same style of macro test as the other architectures?  "|| defined RISCV64"

File: src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp [hunk 2]
```diff
@@ -524,6 +551,37 @@ void ShenandoahBarrierSetAssembler::cmpxchg_oop(MacroAssembler* masm,
...
+
+  // number of bytes to copy
+  __ sub(count, end, start);
+
+  __ load_byte_map_base(tmp);
+  __ add(start, start, tmp);
+
+  __ bind(L_loop);
+  __ add(tmp, start, count);
+  __ sb(zr, Address(tmp));
```
Comment (kdnilsen) [inline src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp:579]: Do we want to condition behavior on UseCondCardMark?  Other architectures are considering this.

**Comment by RealFYang** [comment]
> Could you please enable the github actions for testing in your fork of this repo?

Done. Thanks for reminding.

File: src/hotspot/share/gc/shenandoah/mode/shenandoahGenerationalMode.cpp (hunk 1 above)
Comment (RealFYang) [reply src/hotspot/share/gc/shenandoah/mode/shenandoahGenerationalMode.cpp:36]: Good catch. I think it's mis-typed. Fixed. Thank you.

File: src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp (hunk 2 above)
Comment (RealFYang) [reply src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp:579]: Seems still under discussion? I will provide further update once it's resolved.

**Review by kdnilsen — APPROVED** [review]

File: src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp (hunk 2 above)
Comment (kdnilsen) [reply src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp:579]: That's fine.  It is an optimization that might perform better when/if implemented.

**Review by kdnilsen — APPROVED** [review]

**Comment by RealFYang** [comment]
Thanks all for the review. Merged master and pushed one small improvement for `ShenandoahBarrierSetAssembler::store_check`. This saves us one add instruction when doing address calculation under `UseCondCardMark`. `hotspot_gc_shenandoah` & `gtest` tested with release and fastdebug builds. I am going to integrate this change if no other opinions.

**Pull request closed without merging**